### DISCUSSION TITLE
External user handling with oAuth

### DISCUSSION THREAD

**Discussion by pidal** [directus/directus#5379] [Archive]
Hello,

Im using Directus 9 and managed to configure properly auth0.com as a login provider. 

My question is the following: Is it possible to handle all users with Auth0? At the moment I can only login with an oauth provider with users already registered on the directus DB.

What I'm trying to build is some kind of middleware on directus which is able to check that the JWT on the authorization request (generated by a third party) is valid, and login/register the user accordingly.

This way my users will be able to login from the frontend (angular) with the auth0 provider and directus will be able to check permissions and handle the requests.

Is this a good practice or im all wrong?

Thank you for your time!

**Comment by rijkvanzanten** [comment]
There's currently no native way to register users from "outside" (/ SSO), but there's an in-works feature request here: https://github.com/directus/directus/discussions/3083 / PR here: https://github.com/directus/directus/pull/3955 🙂 

> What I'm trying to build is some kind of middleware on directus which is able to check that the JWT on the authorization request (generated by a third party) is valid, and login/register the user accordingly.

This should totally work! We just landed an `oauth` hook (https://github.com/directus/directus/pull/5045) that you can use to automatically register your user in Directus if it doesn't exist yet, and then make it go through the regular flow 👍🏻

**Reply by pidal** [reply]
Thank you @rijkvanzanten ,

Sorry for my late response. I've been trying to make it work but I cannot access any user data from the hook. I've tried to console.log all the objects passed to the hook: item, accountability, payload, etc... but get undefined for all of them.

My code:

```js
module.exports = function registerHook() {

    return {
        'oauth.*.login': async function ({ event, action, item, accountability, payload }) {
            console.log(event);
            console.log(action);
            console.log(item);
            console.log(accountability);
            console.log(payload);
            console.log("req");
        },
    };

};
```


and the terminal output:

`00:41:09 ✨ request completed GET 302 /auth/oauth/auth0?redirect=http://localhost:8055/admin/login 299ms
undefined
undefined
undefined
undefined
undefined
req
00:41:15 ✨ request completed GET 302 /auth/oauth/auth0/callback?code=QYIeLADqpcclve3D 1s
00:41:15 ✨ request completed GET 200 /admin/login 2ms`

I'm missing something?

Thank you again for your time!

**Reply by rijkvanzanten** [reply]
@pidal The first parameter is the user profile data coming from the oAuth provider, what you're looking for exists as the second parameter 🙂 

https://github.com/directus/directus/blob/1579108dcaff469918d14a3edc72b86b77390da5/api/src/controllers/auth.ts#L297-L305

Try:

```js
module.exports = function registerHook() {

    return {
        'oauth.*.login': async function (oauthProfile, { event, action, item, accountability, payload }) {
            console.log(oauthProfile);
            console.log(event);
            console.log(action);
            console.log(item);
            console.log(accountability);
            console.log(payload);
            console.log("req");
        },
    };

};
```

**Reply by pidal** [reply]
Hi @rijkvanzanten,

Thank you! Now I'm able to see all the oauth provider profile information for the user login in :)

Although im stuck again haha. My problem is that I cant use the services object to check the user and/or create it.

This is my code at the moment:

```
module.exports = function registerHook({ services, database, getSchema }) {
    const { UsersService } = services;

    return {
        'oauth.*.login': async function (oauthProfile, { event, action, item, accountability, payload, schema }) {

            const userService = new UsersService('directus_users', { schema: getSchema, accountability: accountability, knex: database });
            userService
                .readByQuery({ sort: 'email', fields: ['*'] })
                .then((results) => res.json(results))
                .catch((error) => {
                    console.log(error)
                })
```

I get the following error:

`TypeError: Cannot read property 'collections' of undefined`

I've been playing around but didn't managed to accomplish it. Sorry to bother you again, i've checked the documentation but didn't find an answer.

**Reply by rijkvanzanten** [reply]
Very close! The UsersService doesn't need you to specify the collection (as it's just users), so it's `new UsersService({ schema: getSchema, etc` instead of `new UsersService('directus_users', etc)` 👍🏻

**Reply by peteruithoven** [reply]
@pidal did you ever get to finish this. I think I'm trying to do the same thing but I'm also running into issues.  
Biggest issue for me is that I can't find documentation on how to use this UsersServices.

**Reply by pidal** [reply]
Hi @peteruithoven ,

Still struggling with it! I don't have enough time to work on it at the moment (it's a side project) but hope i'll get back at it soon!

@rijkvanzanten instructions were helpful and I managed to intercept the request, but couldn't read/write on the db.

If you manage to make it work please share :) 

Greetings,
Alfonso

**Reply by peteruithoven** [reply]
Bit of progress, I've been able to create the user, but I can't login yet. I think I need to add some kind of oauth kind of info to the account, like you'd normally add password info. 
``` js
module.exports = function registerHook({ services, database, getSchema }) {
  const { UsersService } = services;
  return {
    // based on https://github.com/directus/directus/discussions/5379
    "oauth.*.login": async function (oauthProfile) {
      const { email } = oauthProfile.profile;
      const schema = await getSchema();
      const usersService = new UsersService({
        schema,
        knex: database
      });

      const user = await usersService.createOne({
        email: email,
        role: "1f7e3cdd-b61c-4fb7-8c5a-4c509d52d41d"
      });
      console.log("  user: ", user);
    }
  };
};
```

**Reply by peteruithoven** [reply]
Then again, in this PR I see them also only adding `email` and `status`:
https://github.com/directus/directus/pull/3955/files#diff-2e0966802490ef6b64754924c1f0da2807f05359551e77bf115ce40384d56e0fR267

**Reply by peteruithoven** [reply]
Ah it did work, just not the first time. Trick was to use `.before`. There is also an error thrown when the email is already in use. There is probably a good way to check this, but wrapping the create function in a try catch also works.

``` js
module.exports = function registerHook({ services, database, getSchema }) {
  const { UsersService } = services;
  return {
    // based on https://github.com/directus/directus/discussions/5379
    "oauth.*.login.before": async function (oauthProfile) {
      const { email } = oauthProfile.profile;
      const schema = await getSchema();
      const usersService = new UsersService({
        schema,
        knex: database
      });

      // const accountExists = await usersService.emailHasAccount(email); // Doesn't exist yet...
      // if (accountExists) return;
      try {
        const user = await usersService.createOne({
          email: email,
          role: "1f7e3cdd-b61c-4fb7-8c5a-4c509d52d41d"
        });
        console.log("  created user: ", user);
      } catch (error) {
        console.log("error: ", error);
      }
    }
  };
};
```

**Reply by peteruithoven** [reply]
Added domain whitelisting and a better user existence check, inspired by:
https://github.com/directus/directus/pull/3955

``` js
const whitelistedDomains = ["..."];

module.exports = function registerHook({ services, database, getSchema }) {
  const { UsersService } = services;
  return {
    "oauth.*.login.before": async function (oauthProfile) {
      const { email } = oauthProfile.profile;
      const schema = await getSchema();
      const usersService = new UsersService({
        schema,
        knex: database
      });

      const emailDomainIsWhitelisted = whitelistedDomains.some((d) =>
        email.includes(d)
      );
      if (!emailDomainIsWhitelisted) return;

      const existingUser = await database("directus_users")
        .where({ email })
        .first();
      if (existingUser) return;

      await usersService.createOne({
        email: email,
        role: "1f7e3cdd-b61c-4fb7-8c5a-4c509d52d41d"
      });
    }
  };
};
```
